### ISSUE TITLE
Add update groups, and update before/after for entity systems

### ISSUE THREAD

**Issue by d44m1n** [OurMachinery/themachinery-public#50]
Sometimes it will be useful to update system in specific stage, or before/after specific system.

a rough idea:
```c
enum tm_enitity_update_group {
    // For example `Spin Component`.
    TM_ENITITY_UPDATE_GROUP_SIMULATION,
    // For example `Tessellated Plane Component`.
    TM_ENITITY_UPDATE_GROUP_RENDER
};

**Comment by niklas-ourmachinery** [comment]
Agreed, some way of controlling the update order is needed. We are thinking about how to do this in a way that works well with the very flexible nature of the engine. Having fixed categories `SIMULATION`, `RENDER` etc is problematic, because what if the user has a need to insert an update step between `SIMULATION` and `RENDER`.

We could just let the engine specify an *update_order* `double` and sort the updates by that. This would let you insert an update between any two existing updates. It can become slightly chaotic though, since there won't be any central management of how these values are assigned. For example, two engines from different plugins might get the same *update_order*.

So I'm still mulling over whether there is a better way to do this, which is why I haven't implemented it yet. But my guess is that we'll probably end up with the *update_order* double.

**Comment by d44m1n** [comment]
Yeah, that's a good point.
But there is a problem. No one will have control on `update_order`. 
So for example:
One creator uses range from 0.0 to 1.0, and sets `update_order` to 0.3, because someone thought that this is good place to update this engine/system.
Another person uses range from 0.0 to 100.0, and sets `update_order` to 30.0.
Now you want use these 2 plugins. 
If you have access to source code, then you can just edit values, and you are good to go. 
But what when plugin is closed source?
Even if there will be requirement to use range 0.0 to 1.0, it still gonna be a problem. 
Everyone will do whatever they want, so it will cause a mess.

Categories may be less flexible, but they will be better option imho.
Also you can still add category for every stage, so I don't think that this will cause a big problem.

So categories + update before/after like:
```c
const tm_engine_i spin_engine = {
    ...,
    .update_group = TM_ENITITY_UPDATE_GROUP_SIMULATION,
    .update_before = Y_ENGINE_ID,
    .update_after = X_ENGINE_ID,
};
```
will be much better option imho.

**Comment by niklas-ourmachinery** [comment]
The built-in plugins would would come with `update_order` values, so we would have maybe `SIMULATION = 0` and `RENDER = 100` and that would set the baseline for everything else, so it wouldn't be complete chaos -- still might get messy though.

In addition, I think we should expose a `set_update_order()` that could be called to change the `update_order` of engines from other plugins. So each plugin would come with a default `update_order` value, but if you run into trouble with these defaults, you can always force a certain update order for your game.

**Comment by d44m1n** [comment]
Ok, this solves "standardization" problem.

With `set_update_order()` I see one problem. What if few plugins will update `update_order` value for example for `spin_component`? Now you don't know to which value `update_order` is set.
Changing value by `set_update_order()` will mess up other systems(in this case they already are messed up). 
It will be easier to have control, but this still gonna be mess. 

Also you need to check other systems `update_order` value to be sure that your systems/engines will run in correct order. 

With categories this is much simpler, and less bug-prone. You only specify category, update after/before if needed and you are done. And explicit dependencies between engines/system are good thing, I think . 
The only problems I see with this approach is that user may create circular dependency between systems/engines, or create combinations that make no sense.

**Comment by niklas-ourmachinery** [comment]
Yes if every plugin called `set_update_order()` on everybody else the result would be pretty chaotic. The idea is that this would be a "last resort" to fix ordering problems. It is not something that you would expect every plugin to do. Rather it is something that would only be done in the main gameplay plugin.

Specifying ordering between plugins (i.e. I want to run BEFORE this and that plugin, but AFTER these plugins) would work too. It has the advantage of being a bit more flexible than specifying a hard number and the disadvantage that it is possible to create "impossible" situation where two plugins both want to run BEFORE one another.

It might be preferable, I'll give it some more thought before implementing this.
